### PR TITLE
nektarpp: new port

### PR DESCRIPTION
**Pull request by mdave**
#### Description

Adds a new port for the [Nektar++ spectral element framework](https://www.nektar.info/).

###### Tested on

macOS 10.15.5 19F96
Xcode 11.3.1 11C504

###### Verification 
Have you

- [X] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [X] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [X] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [X] checked your Portfile with `port lint`?
- [X] tried a full install with `sudo port -vst install`?
- [X] tested basic functionality of all binary files?

### CODE REVIEW

File: science/nektarpp/Portfile
Comment (mf2k) [inline science/nektarpp/Portfile:118]: Don't use `cd` in the system command. Instead use the `-W` flag to set the directory:

`system -W {dir} {command}
`

File: science/nektarpp/Portfile
Comment (mf2k) [inline science/nektarpp/Portfile:115]: We prefer using native Tcl commands to `copy` files instead of `system` here. See https://guide.macports.org/#reference.tcl-extensions.

File: science/nektarpp/Portfile
Comment (mf2k) [inline science/nektarpp/Portfile:85]: Unless you have a specific reason to keep the older pythons, only use python 37 and 38 here.

**Comment by mdave** [comment]
@mf2k Thanks for the review! I've pushed those changes and squashed commit.

**Comment by mdave** [comment]
Just thought I'd quickly bump this one in case there is something else you need from my side?

**Comment by ryandesign** [comment]
I've made several additional commits to fix a few different issues. We'll squash everything down to 1 commit when we merge it but I'm leaving them separate for now so that you can see what I changed and why.

I didn't understand your comment:

> ```
> # Require a separate destroot, since there is a conflict between CMake install prefix
> # and the requirement for DESTDIR.
> ```

Replacing the entire destroot phase didn't seem to be necessary so I changed it to just add the python installing instructions.

Was your comment intended to refer to the bug that some files get installed in a weird directory in ${destroot}${portdbpath} and then have to be removed? You should file a bug report about that with the developers (I searched their issue tracker for "DESTDIR" and didn't find any hits) and then we can put the URL into the port as well.

There's still the issue that, when building, I see this:

```
[  1%] Performing download step (download, verify and extract) for 'triangle-1.6'
cd /opt/local/var/macports/build/_Users_rschmidt_macports_macports-ports-ryandesign-fork_science_nektarpp/nektarpp/work/nektar-5.0.0/ThirdParty && /opt/local/bin/cmake -P /opt/local/var/macports/build/_Users_rschmidt_macports_macports-ports-ryandesign-fork_science_nektarpp/nektarpp/work/build/ThirdParty/stamp/download-triangle-1.6.cmake
-- Downloading...
   dst='/opt/local/var/macports/build/_Users_rschmidt_macports_macports-ports-ryandesign-fork_science_nektarpp/nektarpp/work/nektar-5.0.0/ThirdParty/triangle.zip'
   timeout='none'
-- Using src='http://www.nektar.info/thirdparty/triangle.zip'
-- [download 100% complete]
-- [download 0% complete]
-- [download 13% complete]
-- [download 25% complete]
-- [download 38% complete]
-- [download 50% complete]
-- [download 63% complete]
-- [download 75% complete]
-- [download 88% complete]
-- [download 100% complete]
-- verifying file...
       file='/opt/local/var/macports/build/_Users_rschmidt_macports_macports-ports-ryandesign-fork_science_nektarpp/nektarpp/work/nektar-5.0.0/ThirdParty/triangle.zip'
-- Downloading... done
cd /opt/local/var/macports/build/_Users_rschmidt_macports_macports-ports-ryandesign-fork_science_nektarpp/nektarpp/work/nektar-5.0.0/ThirdParty && /opt/local/bin/cmake -P /opt/local/var/macports/build/_Users_rschmidt_macports_macports-ports-ryandesign-fork_science_nektarpp/nektarpp/work/build/ThirdParty/stamp/verify-triangle-1.6.cmake
cd /opt/local/var/macports/build/_Users_rschmidt_macports_macports-ports-ryandesign-fork_science_nektarpp/nektarpp/work/nektar-5.0.0/ThirdParty && /opt/local/bin/cmake -P /opt/local/var/macports/build/_Users_rschmidt_macports_macports-ports-ryandesign-fork_science_nektarpp/nektarpp/work/build/ThirdParty/stamp/extract-triangle-1.6.cmake
-- extracting...
     src='/opt/local/var/macports/build/_Users_rschmidt_macports_macports-ports-ryandesign-fork_science_nektarpp/nektarpp/work/nektar-5.0.0/ThirdParty/triangle.zip'
     dst='/opt/local/var/macports/build/_Users_rschmidt_macports_macports-ports-ryandesign-fork_science_nektarpp/nektarpp/work/nektar-5.0.0/ThirdParty/triangle-1.6'
-- extracting... [tar xfz]
-- extracting... [analysis]
-- extracting... [rename]
-- extracting... [clean up]
-- extracting... done
```

And similar for tetgen.zip (tetgen-1.5). Ports shouldn't download things during the build phase. Any downloading should be done by MacPorts in the fetch phase. Extracting should happen in the extract phase but if that's not possible it would still be an improvement for MacPorts to download the file and put it where the build expects it.

The files it's downloading during the build phase don't have the version number in the name. When they're added as distfiles, we'll want to find versions of these files that do have the version in the name.

You made your PR from the master of your fork. This means you shouldn't update your master until this PR is dealt with. Then you can try to get your master back to a state of mirroring ours if you know how (I don't) or you can delete your fork and make a new one. When you submit PRs in the future, make sure they're from a branch created just for that PR so that you can continue to do other work in your fork while the PR is in progress.

**Comment by ryandesign** [comment]
> Ports shouldn't download things during the build phase.

Did you check whether the separate distfiles upstream offers might already contain these files so that separate downloads would not be necessary? For example this file:

http://www.nektar.info/downloads/file/nektar-source-tar-xz-4/nektar-5.0.0.tar.xz

If that doesn't contain the additional files either, you should file that as a bug report with the developers. They should offer a complete source code download.

We should use that file in any case since if nothing else it is smaller due to its xz compression.

**Comment by mdave** [comment]
Thanks for the comments and patches @ryandesign - that's very helpful in understanding bits I did wrong. Some feedback from my side:

- Note I am one of the developers in this case so this makes things a bit easier! 
- On the issue with `DESTDIR`: this turned out to be a relatively simple patch that simplifies things quite a bit from the Portfile perspective. (CMake was INSTALL'ing the NekPy libraries to a temporary location in the build directory, so calling destroot caused temporary files from the build tree to be installed in the destroot.) I will move that patch back upstream for future releases.
- I've left the source to pull from gitlab, since this will likely be the primary release route for our source going forward and the direct links from website may be removed in the near-ish future. (We are moving towards a continuous delivery model to try and reduce our time between releases.)
- I've modified the Portfile to download the two third-party libraries from the correct location, and have changed them to include version names. When MPI is enabled there is also the `gsmpi` library; I think now that would be everything however.
- We don't include a full tarball release, since we have a number of fairly large dependencies (including boost, OpenCASCADE) which would balloon the full source rather a lot. Many of those are also optional, so we prefer not to do this and rely therefore on CMake ExternalProject to automatically download and compile as required, depending on system dependencies/compile options.

Let me know what you think of these changes and I'm happy to modify further as required.

**Comment by mdave** [comment]
Just bumping this one again before it falls off the radar. Happy to apply further changes as required.

File: science/nektarpp/Portfile
Comment (mf2k) [inline science/nektarpp/Portfile:79]: It appears that there are duplicates in here. Can you sort them alphabetically and remove the dups?

Comment (mdave) [reply science/nektarpp/Portfile]: Okay have done!

File: science/nektarpp/Portfile
Comment (ryandesign) [inline science/nektarpp/Portfile:126]: To avoid repeating version numbers here, you could programmatically copy all files except the main file:

```
    foreach distfile ${distfiles} {
        set distfile [lindex [split ${distfile} :] 0]
        if {${distfile} ne ${main_distfile}} {
            copy ${distpath}/${distfile} ${worksrcpath}/ThirdParty/
        }
    }
```

File: science/nektarpp/Portfile
Comment (mdave) [reply science/nektarpp/Portfile]: Cool, I wondered about this but didn't know whether that sort of functionality existed. I've modified as you've suggested.

File: science/nektarpp/Portfile
Comment (mdave) [reply science/nektarpp/Portfile]: No problem, have changed. Checked lint this time and it doesn't report any issues.

File: science/nektarpp/Portfile
Comment (mdave) [reply science/nektarpp/Portfile:130]: Seems to work well, thanks!

**Comment by mdave** [comment]
@mf2k @ryandesign Have applied the requested changes and marked as resolved; please do let me know if there's anything else you'd like changed!

**Comment by mdave** [comment]
Just thought I would bump this again before it falls off the radar.

File: science/nektarpp/Portfile
Comment (mf2k) [inline science/nektarpp/Portfile:17]: This line is provided by the gitlab portgroup and can be removed.

File: science/nektarpp/Portfile
Comment (mf2k) [inline science/nektarpp/Portfile:9]: These 2 duplicate lines can be removed.

**Review by mf2k — CHANGES_REQUESTED** [review]

File: science/nektarpp/Portfile
Comment (mdave) [reply science/nektarpp/Portfile:17]: Ah okay -- I had added this I think in reply to one of the comments above, but now I can't find it so perhaps I'm imagining things. However, the gitlab instance's homepage is just a generic instance, whereas this line adds the main project webpage. Let me know which you would like to keep but my preference would be the latter.

File: science/nektarpp/Portfile
Comment (mdave) [reply science/nektarpp/Portfile:9]: I'm not sure I see duplicate lines?

Comment (mf2k) [reply science/nektarpp/Portfile:9]: Line 9 is the same as line 17. Line 9 should be removed. 

Line 8 will be overridden by the `gitlab.setup` command (line 12). It will need to be moved to after that line to change the name.

File: science/nektarpp/Portfile
Comment (mf2k) [reply science/nektarpp/Portfile:17]: Sorry, yes this homepage line (line 17) is fine. The duplicate (line 9) should be removed however.

**Review by mf2k — CHANGES_REQUESTED** [review]

File: science/nektarpp/Portfile
Comment (mdave) [reply science/nektarpp/Portfile]: Ah okay great, I totally didn't see the one on line 17. Have removed the one on line 9 now.

File: science/nektarpp/Portfile
Comment (mdave) [reply science/nektarpp/Portfile:16]: Done!

**Comment by mdave** [comment]
Hi both, anything else needed for this?

**Review by mf2k — APPROVED** [review]

File: science/nektarpp/Portfile
Comment (mf2k) [inline science/nektarpp/Portfile:87]: Does it really take such a long time to build the demos that a variant is needed and this is not always done?

**Review by mf2k — CHANGES_REQUESTED** [review]

File: science/nektarpp/Portfile
Comment (mdave) [reply science/nektarpp/Portfile:87]: No, not really -- it just produces rather a lot of additional executables that aren't really all that interesting unless you specifically want them.

Comment (mdave) [reply science/nektarpp/Portfile]: @mf2k I decided this is not really interesting enough to warrant an entire variant, so I've removed it. The demos are quite specialised, for the most part, so if necessary one can compile the demo from source against the library in any case. Let me know if you want any other changes.

File: science/nektarpp/Portfile
Comment (mf2k) [inline science/nektarpp/Portfile:8]: The `name` line needs to be moved to just below the `gitlab.setup` or else it will get overridden.

**Review by mf2k — CHANGES_REQUESTED** [review]
One more tweak

File: science/nektarpp/Portfile
Comment (mdave) [reply science/nektarpp/Portfile]: @mf2k No problem, have moved.

**Review by mf2k — APPROVED** [review]

**Comment by mdave** [comment]
@mf2k Not sure what the next steps are here -- do you want me to rebase against current master & squash commits?

**Comment by mf2k** [comment]
Yes definitely. I was hoping for another set of eyes on this one.

**Comment by mdave** [comment]
@mf2k No worries, have rebased & squashed.

**Comment by mf2k** [comment]
Hmmm. No other approvals. @mdave : Is this still the latest version? Can you update it if not? Otherwise let me know and I will commit it.

**Comment by mdave** [comment]
@mf2k No newer version yet -- we should have a patch release in the early new year which ports these patches upstream, so if you are happy to merge this version I can update once that lands.

**Pull request merged**
